pubtools/exodus/hooks/pulp.py

```diff
@@ -0,0 +1,165 @@
+
+        if not self._gw_key:
+            LOG.debug("EXODUS_GW_KEY not found, authentication may fail")
+
+        auth_url = urljoin(self._gw_url, "/whoami")
+        resp = self._do_request(method="GET", url=auth_url)
+        context = resp.json()
+
+        for user_type, ident in (
+            ("client", "serviceAccountId"),
+            ("user", "internalUsername"),
+        ):
+            typed_ctx = context[user_type]
+            if typed_ctx["authenticated"]:
+                LOG.debug(
+                    "You are authenticated as %s %s with roles: %s",
+                    user_type,
+                    typed_ctx[ident],
+                    [str(role) for role in typed_ctx["roles"]],
+                )
+                break
+        else:
+            LOG.debug("You are not authenticated with exodus-gw.")
+
+    @classmethod
+    def _unpack_response(cls, response):
+        """Raise if 'response' was not successful.
+
+        This is the same as response.raise_for_status(), merely wrapping it
+        to ensure the body is logged when possible."""
+
+        try:
+            response.raise_for_status()
+        except Exception as outer:
+            try:
+                body = response.json()
+            except:
+                raise outer
+
+            LOG.error("Unsuccessful response from exodus-gw: %s", body)
+            raise
+
+    @backoff.on_predicate(
+        wait_gen=backoff.expo,
+        predicate=lambda task: task["state"] not in ("FAILED", "COMPLETE"),
+        max_time=1800,
+    )
+    def _poll_commit_completion(self, commit):
+        """Issues request(s) to exodus-gw to get the task state, returning once
+        it either is FAILED or COMPLETED."""
+        task_url = urljoin(self._gw_url, commit["links"]["self"])
+
+        resp = self._do_request(method="GET", url=task_url)
+        task = resp.json()
+
+        LOG.info("Task state: %s", task["state"])
+
+        return task
+
+    @hookimpl
+    def pulp_repository_pre_publish(self, to_publish):
+        """Invoked before a Pulp repository is published.
+
+        This implementation adds to each config the --exodus-publish argument,
+        attaching the repository to an Exodus gateway publish.
+
+        Args:
+            to_publish (list[tuple[:class:`~pubtools.pulplib.Distributor`, dict]])
+                List of Distributor-config pairs intended for publishing.
+
+        Returns:
+            to_publish (list[tuple[:class:`~pubtools.pulplib.Distributor`, dict]])
+                List of potentially adjusted Distributor-config pairs to publish.
+        """
+
+        if not self._publish:
+            publish_url = os.path.join(self._gw_url, self._gw_env, "publish")
+            resp = self._do_request(method="POST", url=publish_url)
+            self._publish = resp.json()
+            LOG.debug("Created publish: %s", self._publish["id"])
+
+        publish_id = self._publish["id"]
+
+        for (_, config) in to_publish:
```

Comment:
I think this is broken now as to_publish will have entries also for some distributors which don't accept rsync_extra_args (like yum_distributor).  Possibly the distributors will just ignore it, but I think they do generally validate their args.

In any case though, I have suggested on the pubtools-pulplib change to make this higher-level.  If that suggestion is done then I think the code here will look something more like:

```python
args = options.rsync_extra_args or []
args.append("--exodus-publish=%s" % publish_id)
return attr.evolve(options, rsync_extra_args=args)
```

Reply:
This now depends on merge of https://github.com/release-engineering/pubtools-pulplib/pull/156